### ISSUE TITLE
Image data fails to update when a byte is changed directly

### ISSUE THREAD

**Issue by CratedCurse** [godotengine/godot#65870]
### Godot version

4.0.beta1

### System information

Windows 11

### Issue description

When trying to change a byte in the data byte array by index, the byte in the array doesn't seem to change. Running the same code in 3.5.stable works correctly.

Input:
```gdscript
func _ready() -> void:
	var image: Image = Image.new()
	image.create(5, 5, false, Image.FORMAT_R8)
	image.data["data"][0] = 5
	image.data["data"][1] = 10
	image.data["data"][2] = 15
	
	print(image.data["data"])
```

Output:

`[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]`

The output should look like this:

`[5, 10, 15, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]`

### Steps to reproduce

- Create an image with any size, no mipmaps, and format r8.
- Change a byte in the array with any valid index.
- Print the array.

**Comment by gotnospirit** [comment]
`get_data` is a const function, I think this was a bug in 3.x
try
```
	image.set_pixel(0, 0, Color8(5,0,0))
	image.set_pixel(1, 0, Color8(10,0,0))
	image.set_pixel(2, 0, Color8(15,0,0))
```

**Comment by gotnospirit** [comment]
or
```
	var data: PackedByteArray = []
	data.resize(25)
	data[0] = 5
	data[1] = 10
	data[2] = 15

	var image: Image = Image.new()
	image.create_from_data(5, 5, false, Image.FORMAT_R8, data)
```

**Comment by CratedCurse** [comment]
I'm trying to avoid using set_pixel because it's a bit much for what I need, Also I want to keep that image around to update the data later when I need to.

> `get_data` is a const function, I think this was a bug in 3.x try
> 
> ```
> 	image.set_pixel(0, 0, Color8(5,0,0))
> 	image.set_pixel(1, 0, Color8(10,0,0))
> 	image.set_pixel(2, 0, Color8(15,0,0))
> ```
> or
> 
> ```
> 	var data: PackedByteArray = []
> 	data.resize(25)
> 	data[0] = 5
> 	data[1] = 10
> 	data[2] = 15
> 
> 	var image: Image = Image.new()
> 	image.create_from_data(5, 5, false, Image.FORMAT_R8, data)
> ```

**Comment by timothyqiu** [comment]
In `3.x`, `image.data["data"][0] = 5` is more or less equivalent to

```gdscript
var data = image.data
data["data"][0] = 5
image.data = data
```

In `4.x`, the setter is not called. They both make sense so I'm not sure if this is intended. Currently, you can workaround this by storing the dictionary into a variable like in the snippet above.

**Comment by alazifk** [comment]
I think it should be made possible to just do `image.data[0] = 255`. Since locking and unlocking is no longer a thing I see no reason not to allow directly manipulating the data, it makes sense for some use cases.

**Comment by alazifk** [comment]
I suspect performance-wise this may be slower than using `set_pixel()` and `get_pixel()`. I did some experimenting (potentially sloppy, in debug mode) filling a 2048*2048 RGBA texture:
- `Image.fill()` 2ms
- custom `set_pixels()` function 300 ms
- exposing `set_color_at_offset` 2100 ms
- looping over the image with `set_pixel()` 2500 ms
- "direct" access 3800 ms

**Comment by HolonProduction** [comment]
Yeah, GDScript optimizes away the setter call on packed arrays, since in theory they are passed by reference. This is a bit of a tricky one, since this behaves differently for packed arrays in GDScript and core API's:

> Note: Packed arrays are always passed by reference. To get a copy of an array that can be modified independently of the original array, use duplicate(). This is not the case for built-in properties and methods. In these cases the returned packed array is a copy, and changing it will not affect the original value. To update a built-in property of this type, modify the returned array and then assign it to the property again.

In a case like this were it is wrapped into a dictionary as well, I see no way for GDScript to distinguish between both of those cases.
